### ISSUE TITLE
Add ability to store message IDs in option redis queue

### ISSUE THREAD

**Issue by patcon** [42wim/matterbridge#541]
So that between Slack instance, we can preserve threading (https://github.com/42wim/matterbridge/issues/529) and send emojis (https://github.com/42wim/matterbridge/issues/526) with more consistent behaviour :)

Thoughts? I could be pretty simple, and totally optional for people who choose to stay utterly stateless. 
With [`matterbridge-heroku`](https://github.com/patcon/matterbridge-heroku), adding persistence of relationships could be as simple as:

    $ heroku addons:create heroku-redis:hobby-dev

**Comment by 42wim** [comment]
Quick thoughts, redis seems like overkill. (and isn't really made for persistency afaik)
I'm more for using something like boltdb (https://github.com/etcd-io/bbolt) (simple embedded database).

As long as it's optional (and simple) it's ok for me.

**Comment by patcon** [comment]
Bolt seems cool but I'd really prefer something that continues to work easily on heroku, where I have a few instances running :)

I'll think on it. I think you're right about redis not being for persistence. Perhaps postgres. Free plan has 10,000 rows, which exactly matches how much history our slack instance has anyhow...!

**Comment by 42wim** [comment]
Why do you think bolt wouldn't work?

**Comment by patcon** [comment]
Unless I'm misunderstanding, no heroku service addins offer bolt. Herokubis stateless without a file system, sonit needs an add-on provider for its ability to have arrives (databases, caches, etc) as push-button

**Comment by patcon** [comment]
Ok, so on further investigation, redis actually seems a step in the right direction. I don't think we need full persistency perhaps, but we need to be able to query based on more than primary key (original message timestamp), which is only what the lru cache allows.

We instead need something that would allow us to query for a mID value in BrMsgID from the lru cache (using a ThreadTimestamp from thread reply ), and get the key (which is the original Timestamp). We could then know all the other messages to update from there.

**Comment by 42wim** [comment]
We should use https://github.com/abronan/valkeyrie if this gets implemented.

**Comment by patcon** [comment]
perrrrfect. I'll prob work on this next. Getting to crunch time on another project, so might not be as soon as I'd like

**Comment by patcon** [comment]
Potentially helpful link: https://medium.com/@nevio/proper-way-how-to-marshal-and-unmarshal-golang-structs-within-the-redis-cache-855423902787

Was curious about using msgpack (which redis-cache golang lib uses), but seems marshalling as json should work fine: https://www.reddit.com/r/golang/comments/476fdi/what_is_best_serialization_format_for_boltdb_for/

**Comment by 42wim** [comment]
The valkeyrie library should handle all that there won't be any redis or bolt specific code in matterbridge.

**Comment by patcon** [comment]
No, valkeyrie doesn't do marshalling -- just accepts byte slices as far as I can tell. Anyhow, I have a code spike here:
https://github.com/42wim/matterbridge/compare/master...patcon:valkeyrie-persistence?expand=1

Busy week coming up, so that's likely all I'll get done, but anyone is welcome to pull it and fix it up :)

**Comment by sas1024** [comment]
May be use [gokv](https://github.com/philippgille/gokv) - Simple key-value store abstraction and implementations for Go (Redis, Consul, etcd, bbolt, BadgerDB, LevelDB, Memcached, DynamoDB, S3, PostgreSQL, MongoDB, CockroachDB and many more)
?

**Comment by sas1024** [comment]
@patcon hello! Can you share your code with valkeyrie implementation for persistent message store? I want to implement this thing and any help will be useful :)

**Comment by sas1024** [comment]
@42wim hello! Can you share your thoughts on implementing this feature? I want to realize this feature.

For now in code i see:
- HashiCorp LRU cache usage.
- This cache initialized and used in multiple places (in gateway, slack, discord, mattermost, rocketchat packages)

IMHO there is two possible implementations for persistent storing message IDs:
1. Completely replace LRU cache with valkeyrie.
2. Use LRU cache and valkeyrie together. If no caching backend was configured, then will be used LRU cache. In this implementation we need to create abstraction between LRU cache and valkeyrie.

And we can initialize cache once and use it in all required places.

**Comment by sas1024** [comment]
@42wim ping? :)

**Comment by patcon** [comment]
A-ha! I realized that I transferred the repo to another org, and my personal fork was a more recent one without my old branches. The code is _really_ old, but it's here 🎉
https://github.com/42wim/matterbridge/compare/master...patcon:valkeyrie-persistence?expand=1

Anyhow, good luck and godspeed @sas1024 ! I'm rooting for you!

**Comment by celine-m-s** [comment]
Hey there! Any updates on this? I just implemented Matterbridge to synchronize messages between Slack and Mattermost but it tends to crash twice a week. This is not a problem in itself, except it breaks all threads (which is quite annoying) so I'd be very happy if persistent threads were in the roadmap.
By the way, thanks for all the hard work! :clap:

**Comment by yousefmansy1** [comment]
@sas1024 did you ever get anywhere on writing a solution for this feature?
I'd be willing to trying my hand at it as well.

The code link @patcon posted does not appear to exist anymore either, hoping you have it saved somewhere?
> https://github.com/42wim/matterbridge/compare/master...patcon:valkeyrie-persistence?expand=1

**Comment by sas1024** [comment]
@yousefmansy1 i'm sorry, but I haven't started working on this feature yet :(
It well be fantastic, if you will try to realize this feature.

**Comment by sas1024** [comment]
No, I don't have this code, this link was broken several months.
May be @patcon can share this code again?

**Comment by yousefmansy1** [comment]
No worries, thanks for the response.
I'll probably try to take a look using the gokv thing you mentioned.

**Comment by patcon** [comment]
oy... i'm sorry, i'm seeing that I must have confused to clipboard, and shared the same link again. It can't be merged anymore without conflict resolution, but here's the branch with the work:
https://github.com/g0v-network/matterbridge/tree/valkeyrie-persistence
